File: 541_ReverseString_II.py
##90.83%
class Solution(object):
    def reverseStr(self, s, k):
        """
        :type s: str
        :type k: int
        :rtype: str
        """
        if not s or k <= 0:
            return s
        start, step = 0, k * 2
        s = list(s)
        while start < len(s):
            end = start + k if start + k < len(s) else len(s)
            s[start:end] = s[start:end][::-1]
            start += step
        return "".join(s)

##97.9%		
class Solution(object):
    def reverseStr(self, s, k):
        """
        :type s: str
        :type k: int
        :rtype: str
        """
        return s[:k][::-1] + s[k:2*k] + self.reverseStr(s[2*k:], k) if s else ""

File: 541_ReverseString_II.cpp
//69.3%
class Solution {
public:
    string reverseStr(string s, int k) {
		if(s.empty() || k <= 0)
			return s;
		int step(k * 2);
        for(int start = 0; start < s.size(); start += step){
			int pend = start + k  < s.size() ? start + k : s.size();
			int pstart(start);
			reverse(s.begin() + start, s.begin() + pend);
		}
		return s;		
    }
};

class Solution {
public:
    string reverseStr(string s, int k) {
        for (int i = 0; i < s.size(); i += 2*k) reverse(s.begin()+i, min(s.begin()+i+k, s.end()));
        return s;
    }
};